File: internal/merge/merge.go
package merge

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"
)

// Entry represents a single file in a merge operation.
type Entry struct {
	Path          string
	HeadingOffset int
}

// ContentsFile represents a contents.yaml structure.
type ContentsFile struct {
	HeadingOffset int   `yaml:"heading_offset"`
	Files         []any `yaml:"files"`
}

// ParseContentsYAML reads a contents.yaml and returns merge entries.
// Paths are resolved relative to the directory containing the YAML file.
func ParseContentsYAML(yamlPath string) ([]Entry, error) {
	data, err := os.ReadFile(yamlPath)
	if err != nil {
		return nil, fmt.Errorf("无法读取 %s: %w", yamlPath, err)
	}

	var cf ContentsFile
	if err := yaml.Unmarshal(data, &cf); err != nil {
		return nil, fmt.Errorf("解析 %s 失败: %w", yamlPath, err)
	}

	baseDir := filepath.Dir(yamlPath)
	var entries []Entry

	for _, item := range cf.Files {
		switch v := item.(type) {
		case string:
			entries = append(entries, Entry{
				Path:          filepath.Join(baseDir, v),
				HeadingOffset: cf.HeadingOffset,
			})
		case map[string]any:
			p, _ := v["path"].(string)
			if p == "" {
				continue
			}
			offset := cf.HeadingOffset
			if ho, ok := v["heading_offset"]; ok {
				switch n := ho.(type) {
				case int:
					offset = n
				case float64:
					offset = int(n)
				}
			}
			entries = append(entries, Entry{
				Path:          filepath.Join(baseDir, p),
				HeadingOffset: offset,
			})
		}
	}

	return entries, nil
}

// ResolveGlobs expands glob patterns and returns sorted, deduplicated .md file paths.
func ResolveGlobs(patterns []string) ([]string, error) {
	seen := make(map[string]bool)
	var result []string

	for _, pattern := range patterns {
		matches, err := filepath.Glob(pattern)
		if err != nil {
			return nil, fmt.Errorf("glob 模式无效 '%s': %w", pattern, err)
		}
		for _, m := range matches {
			if !strings.HasSuffix(strings.ToLower(m), ".md") {
				continue
			}
			abs, err := filepath.Abs(m)
			if err != nil {
				continue
			}
			if !seen[abs] {
				seen[abs] = true
				result = append(result, m)
			}
		}
	}

	sort.Strings(result)
	return result, nil
}

var headingRe = regexp.MustCompile(`^(#{1,6})\s`)

// OffsetHeadings increases heading levels by offset.
// Headings are capped at level 6.
func OffsetHeadings(content string, offset int) string {
	if offset <= 0 {
		return content
	}

	lines := strings.Split(content, "\n")
	for i, line := range lines {
		m := headingRe.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		level := min(len(m[1])+offset, 6)
		lines[i] = strings.Repeat("#", level) + line[len(m[1]):]
	}
	return strings.Join(lines, "\n")
}

// MergeFiles reads all entries, applies heading offsets, and returns combined markdown.
// The first file's frontmatter is preserved; subsequent files' frontmatter is stripped.
func MergeFiles(entries []Entry) (string, error) {
	if len(entries) == 0 {
		return "", fmt.Errorf("没有要合并的文件")
	}

	var parts []string
	for i, e := range entries {
		data, err := os.ReadFile(e.Path)
		if err != nil {
			return "", fmt.Errorf("无法读取 %s: %w", e.Path, err)
		}
		content := string(data)

		if i == 0 {
			// Keep first file as-is (including frontmatter), then offset body
			fm, body := splitFrontmatter(content)
			body = OffsetHeadings(body, e.HeadingOffset)
			if fm != "" {
				parts = append(parts, fm+"\n"+body)
			} else {
				parts = append(parts, body)
			}
		} else {
			// Strip frontmatter from subsequent files
			_, body := splitFrontmatter(content)
			body = OffsetHeadings(body, e.HeadingOffset)
			parts = append(parts, body)
		}
	}

	return strings.Join(parts, "\n\n") + "\n", nil
}

// splitFrontmatter separates YAML frontmatter from body.
// Returns (frontmatter_block_including_delimiters, body).
func splitFrontmatter(content string) (string, string) {
	if !strings.HasPrefix(content, "---") {
		return "", content
	}

	rest := content[3:]
	idx := strings.Index(rest, "\n---")
	if idx < 0 {
		return "", content
	}

	fmEnd := 3 + idx + 4 // "---" + rest[:idx] + "\n---"
	fm := content[:fmEnd]
	body := content[fmEnd:]
	if len(body) > 0 && body[0] == '\n' {
		body = body[1:]
	}
	return fm, body
}

// IsContentsYAML checks if a path looks like a contents.yaml file.
func IsContentsYAML(path string) bool {
	ext := strings.ToLower(filepath.Ext(path))
	return ext == ".yaml" || ext == ".yml"
}

File: src/md2docx/merge.py
"""Multi-file merge: combine multiple markdown files into one before conversion."""

from __future__ import annotations

import glob as glob_mod
from dataclasses import dataclass
from pathlib import Path

import yaml


@dataclass(frozen=True)
class MergeEntry:
    path: Path
    heading_offset: int = 0


def parse_contents_yaml(contents_path: Path) -> list[MergeEntry]:
    """Parse a contents.yaml file into a list of MergeEntry."""
    raw = yaml.safe_load(contents_path.read_text(encoding="utf-8"))
    if not isinstance(raw, dict) or "files" not in raw:
        raise ValueError("contents.yaml 格式错误: 需要顶层 'files' 键")

    base_dir = contents_path.parent
    global_offset = raw.get("heading_offset", 0)
    entries: list[MergeEntry] = []

    for item in raw["files"]:
        if isinstance(item, str):
            path = base_dir / item
            entries.append(MergeEntry(path=path, heading_offset=global_offset))
        elif isinstance(item, dict):
            file_path = base_dir / item["path"]
            offset = item.get("heading_offset", global_offset)
            entries.append(MergeEntry(path=file_path, heading_offset=offset))
        else:
            raise ValueError(f"contents.yaml 中的文件条目格式错误: {item}")

    return entries


def resolve_glob_patterns(patterns: list[str], base_dir: Path) -> list[MergeEntry]:
    """Resolve glob patterns into MergeEntry list."""
    entries: list[MergeEntry] = []
    seen: set[Path] = set()

    for pattern in patterns:
        if base_dir and not Path(pattern).is_absolute():
            full_pattern = str(base_dir / pattern)
        else:
            full_pattern = pattern

        matches = sorted(glob_mod.glob(full_pattern))
        for match in matches:
            p = Path(match).resolve()
            if p not in seen and p.suffix == ".md":
                seen.add(p)
                entries.append(MergeEntry(path=p))

    return entries


def _offset_headings(text: str, offset: int) -> str:
    """Add heading level offset by prepending '#' characters."""
    if offset <= 0:
        return text

    lines = text.split("\n")
    result = []
    for line in lines:
        stripped = line.lstrip()
        if stripped.startswith("#"):
            # Count existing heading level
            hashes = 0
            for ch in stripped:
                if ch == "#":
                    hashes += 1
                else:
                    break
            if hashes <= 6 and (len(stripped) == hashes or stripped[hashes] == " "):
                new_level = min(hashes + offset, 6)
                result.append("#" * new_level + stripped[hashes:])
                continue
        result.append(line)

    return "\n".join(result)


def merge_files(entries: list[MergeEntry]) -> str:
    """Merge multiple markdown files into a single markdown string."""
    parts: list[str] = []
    missing: list[str] = []

    for entry in entries:
        if not entry.path.exists():
            missing.append(str(entry.path))
            continue

        content = entry.path.read_text(encoding="utf-8").rstrip()
        if entry.heading_offset > 0:
            content = _offset_headings(content, entry.heading_offset)
        parts.append(content)

    if missing:
        raise FileNotFoundError("以下文件不存在:\n" + "\n".join(f"  - {m}" for m in missing))

    return "\n\n".join(parts) + "\n"
